plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -264,10 +247,119 @@ ClusterState deleteQueryGroupInClusterState(final String name, final ClusterStat
+            }
+
+            @Override
+            public ThrottlingKey getClusterManagerThrottlingKey() {
+                return updateQueryGroupThrottlingKey;
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                logger.warn("Failed to update QueryGroup due to error: {}, for source: {}", e.getMessage(), source);
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                String name = toUpdateGroup.getName();
+                Optional<QueryGroup> findUpdatedGroup = newState.metadata()
+                    .queryGroups()
+                    .values()
+                    .stream()
+                    .filter(group -> group.getName().equals(name))
+                    .findFirst();
+                assert findUpdatedGroup.isPresent();
+                QueryGroup updatedGroup = findUpdatedGroup.get();
+                UpdateQueryGroupResponse response = new UpdateQueryGroupResponse(updatedGroup, RestStatus.OK);
+                listener.onResponse(response);
+            }
+        });
+    }
+
+    /**
+     * Modify cluster state to update the existing QueryGroup
+     * @param updateQueryGroupRequest {@link QueryGroup} - the QueryGroup that we want to update
+     * @param currentState - current cluster state
+     */
+    ClusterState updateQueryGroupInClusterState(UpdateQueryGroupRequest updateQueryGroupRequest, ClusterState currentState) {
+        final Metadata metadata = currentState.metadata();
+        final Map<String, QueryGroup> existingGroups = currentState.metadata().queryGroups();
+        String name = updateQueryGroupRequest.getName();
+
+        final QueryGroup existingGroup = existingGroups.values()
+            .stream()
+            .filter(group -> group.getName().equals(name))
+            .findFirst()
+            .orElseThrow(() -> new ResourceNotFoundException("No QueryGroup exists with the provided name: " + name));
+
+        // build the QueryGroup with updated fields
+        final Map<ResourceType, Double> updatedResourceLimits = new HashMap<>(existingGroup.getResourceLimits());
+        if (updateQueryGroupRequest.getResourceLimits() != null && !updateQueryGroupRequest.getResourceLimits().isEmpty()) {
+            validateTotalUsage(existingGroups, name, updateQueryGroupRequest.getResourceLimits());
+            updatedResourceLimits.putAll(updateQueryGroupRequest.getResourceLimits());
+        }
+
+        final ResiliencyMode mode = Optional.ofNullable(updateQueryGroupRequest.getResiliencyMode())
+            .orElse(existingGroup.getResiliencyMode());
+
+        final QueryGroup updatedGroup = new QueryGroup(
+            name,
+            existingGroup.get_id(),
+            new ChangeableQueryGroup(mode, updatedResourceLimits),
+            Instant.now().getMillis()
+        );
```

Comment:
This logic of merging the existing and update group should be delegated to `QueryGroup/ChangeableQueryGroup`. Validation can be done after creation of updated query group